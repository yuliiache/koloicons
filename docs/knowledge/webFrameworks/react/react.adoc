== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







