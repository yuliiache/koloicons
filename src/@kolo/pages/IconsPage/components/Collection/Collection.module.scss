@import "@kolo/design";
.collection {
  height: calc(100% - 54px);
}
.collectionIconList {
  padding: 16px;
  display: flex;
  flex-wrap: wrap;
}
.signIn {
  height: 100%;
  padding: 16px 0;
  margin-top: -50px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin-bottom: 20px;
}
.upIcon {
  fill: $WhiteBG;
}
.downloadOptionButton {
  position: fixed;
  z-index: 15;
  bottom: 0;
  border-right: none;
  width: 344px;
}
.positionAbsolute {
  position: absolute;
}
