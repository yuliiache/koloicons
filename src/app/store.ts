import {getReducers} from '@kolo/core/reducers/getReducers';
import {rootSagas} from '@kolo/core/sagas/rootSagas';
import {configureStore} from '@reduxjs/toolkit';
import createSagaMiddleware from 'redux-saga';

const sagaMiddleware = createSagaMiddleware();

const store = configureStore({
  reducer: getReducers(),
  devTools: process.env.NODE_ENV !== 'production',
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(sagaMiddleware),
});

sagaMiddleware.run(rootSagas);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
