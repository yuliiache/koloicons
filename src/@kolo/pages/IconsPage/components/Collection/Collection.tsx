import {isAuthenticatedSelector} from '@kolo/pages/LoginPage/selectors';
import ButtonWithIcon from '@kolo/uiKit/Button/ButtonWithIcon/ButtonWithIcon';
import {ButtonColor, ButtonContentPlacement, ButtonSize} from '@kolo/uiKit/Button/constants';
import Icon from '@kolo/uiKit/Icons';
import SignInMessage from '@kolo/uiKit/SignInMessage/SignInMessage';
import {TooltipPlacement} from '@kolo/uiKit/Tooltip/constants';
import Tooltip from '@kolo/uiKit/Tooltip/Tooltip';
import {useAppSelector} from 'app/hooks';
import classnames from 'classnames';
import {FC, useState} from 'react';

import styles from './Collection.module.scss';
import CollectionIconCard from './components/CollectionIconCard';
import {useScrollPositionTrigger} from './hooks/useScrollPositionTrigger';

const iconData = {
  //TODO: Delete dummy data after getting data from store
  category: {
    name: 'Financial',
    number: 7,
    id: '5d45660c241234001664820c',
  },
  name: 'Mobile payment',
  size: '60px',
  style: 'line',
  id: '5d6a87bdef38414cb068e989',
  isFree: false,
};
const dummyList = Array(4).fill(iconData);

const Collection: FC = () => {
  const collectionIconList = dummyList; //TODO: Get CollectionIconList from store
  const isAuthenticated = useAppSelector(isAuthenticatedSelector);
  const [numberAddedToCollection] = useState(4); //TODO: Get numberAddedToCollection = CollectionIconList.length from store
  const isAbsolutePosition = useScrollPositionTrigger();
  const downloadOptionButtonStyles = classnames(styles.downloadOptionButton, {
    [styles.positionAbsolute]: isAbsolutePosition,
  });
  let tooltipLabel = '';

  if (isAuthenticated && numberAddedToCollection === 0) {
    tooltipLabel = 'Add an icon to Collection';
  } else if (!isAuthenticated) {
    tooltipLabel = 'Only for registered users';
  }

  return (
    <div className={styles.collection}>
      {!isAuthenticated && (
        <div className={styles.signIn}>
          <SignInMessage title="Collections" />
        </div>
      )}
      {isAuthenticated && (
        <>
          <div className={styles.collectionIconList}>
            {collectionIconList.map((iconData, index) => (
              <CollectionIconCard
                key={`${iconData.name}${index}`} // TODO:Delete index from key after adding collection selector
                icon={{
                  name: iconData.name,
                  style: iconData.style,
                  size: iconData.size,
                  id: iconData.id,
                }}
                categoryNumber={iconData.category.number}
                categoryName={iconData.category.name}
                isFree={iconData.isFree}
              />
            ))}
          </div>
        </>
      )}
      <Tooltip
        id={tooltipLabel}
        label={tooltipLabel}
        placement={TooltipPlacement.TOP}
      >
        <div className={downloadOptionButtonStyles}>
          <ButtonWithIcon
            text="Download options"
            isDisabled={numberAddedToCollection > 0 ? false : true}
            icon={<Icon.Up className={styles.upIcon} />}
            size={ButtonSize.FULL_WIDTH}
            placement={ButtonContentPlacement.RIGHT}
            color={ButtonColor.PRIMARY}
            isDownloadOptionButton
          />
        </div>
      </Tooltip>
    </div>
  );
};

export default Collection;
