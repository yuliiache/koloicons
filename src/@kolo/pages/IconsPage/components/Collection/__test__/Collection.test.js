import React from 'react';
import {Provider} from 'react-redux';
import {MemoryRouter} from 'react-router-dom';
import {render} from '@testing-library/react';
import store from 'app/store';

import Collection from '../Collection';

describe('Collection component', () => {
  it('renders correctly and matches snapshot', () => {
    const {asFragment} = render(
      <MemoryRouter>
        <Provider store={store}>
          <Collection />
        </Provider>
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
