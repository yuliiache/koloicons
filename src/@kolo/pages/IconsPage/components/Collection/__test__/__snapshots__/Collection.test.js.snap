// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="collection"
  >
    <div
      class="signIn"
    >
      <div
        class="signInMessage"
      >
        <p
          class="signInText"
        >
          Sign In to work with
          <span
            class="signInTitle"
            data-testid="signInMessageComponent"
          >
             Collections
          </span>
        </p>
        <a
          data-testid="buttonLinkComponent"
          href="/login"
        >
          <button
            class="button button-full-width button-basic button-center"
            data-testid="buttonComponent"
            type="button"
          >
            <span
              data-testid="buttonComponentText"
            >
              Sign In
            </span>
          </button>
        </a>
      </div>
    </div>
    <div
      class="downloadOptionButton"
    >
      <button
        class="button button-full-width button-primary button-right button-download-options"
        type="button"
      >
        <div
          class="buttonContent buttonReverseOrder"
        >
          <span>
            <svg
              class="upIcon"
              data-testid="icon-up"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13.7071 10.7071C13.3466 11.0676 12.7794 11.0953 12.3871 10.7903L12.2929 10.7071L8 6.415L3.70711 10.7071C3.34662 11.0676 2.77939 11.0953 2.3871 10.7903L2.29289 10.7071C1.93241 10.3466 1.90468 9.77939 2.2097 9.3871L2.29289 9.29289L8 3.58579L13.7071 9.29289C14.0976 9.68342 14.0976 10.3166 13.7071 10.7071Z"
              />
            </svg>
          </span>
          <span>
            Download options
          </span>
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
